### PR TITLE
Remove J/JB/JSparse/JSparseB duplication, fixes #468

### LINKED ISSUE #468: reduce J, JSparse, JB, JSparseB code duplication
J, JSparse, JB, JSparseB all contain the same entries but are currently compiled multiple times, which makes up most of the compilation time when compiling from pysb. 

Would be nice to simplify them to a single function such that they only need to compiled once.

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

**Comment by FFroehlich** [comment]
I hope the "only" thing still missing is implementation of sparse blockwise transpose required for second order sensitivities.

**Comment by FFroehlich** [comment]
> I hope the "only" thing still missing is implementation of sparse blockwise transpose required for second order adjoint sensitivities.

**Comment by FFroehlich** [comment]
image

image

File: include/amici/sundials_matrix_wrapper.h [hunk 1]
```diff
@@ -252,6 +252,28 @@ class SUNMatrixWrapper {
                          sunindextype *w, gsl::span<realtype> x,
                          const sunindextype mark,
                          SUNMatrixWrapper *C, sunindextype nnz) const;
+    
+    /**
+     * @brief Computes transpose A' of sparse matrix A and writes is to the matrix C = alpha * A'
```
Comment (dweindl) [inline include/amici/sundials_matrix_wrapper.h:257]:
```suggestion
     * @brief Computes transpose A' of sparse matrix A and writes it to the matrix C = alpha * A'
```

File: include/amici/sundials_matrix_wrapper.h [hunk 2]
```diff
@@ -252,6 +252,28 @@ class SUNMatrixWrapper {
...
+     *
+     * @param C output matrix (sparse or dense)
+     * @param alpha scalar multiplier
+     */
+    void transpose(SUNMatrix C, const realtype alpha) const;
```
Comment (dweindl) [inline include/amici/sundials_matrix_wrapper.h:262]: Don't we want to use `SUNMatrixWrapper`?

File: include/amici/sundials_matrix_wrapper.h [hunk 3]
```diff
@@ -252,6 +252,28 @@ class SUNMatrixWrapper {
...
+    
+    /**
+     * @brief Writes a sparse matrix A to a dense matrix D.
+     *
+     * @param D dense output matrix
+     */
+    void to_dense(SUNMatrix D) const;
```
Comment (dweindl) [inline include/amici/sundials_matrix_wrapper.h:269]: Don't we want to use `SUNMatrixWrapper`?

**Review by dweindl — APPROVED** [review]
Neat cleanup :+1:

File: include/amici/sundials_matrix_wrapper.h (hunk 2 above)
Comment (FFroehlich) [reply include/amici/sundials_matrix_wrapper.h:262]: We use this to directly write to the raw SUNMatrix that comes from sundials, so this needs to be SUNMatrix. Could add additional overload for `SUNMatrixWrapper`, but would currently be unused.

**Pull request merged**